osu.Game/Online/Chat/ChannelManager.cs

```diff
@@ -408,13 +416,51 @@ public void LeaveChannel(Channel channel)
 
             joinedChannels.Remove(channel);
 
+            // Prevent the closedChannel list from exceeding the max size
+            // by removing the oldest element
+            if (closedChannels.Count >= closed_channels_max_size)
+            {
+                closedChannels.RemoveAt(0);
+            }
+
+            // insert at the end of the closedChannels list
+            closedChannels.Insert(closedChannels.Count, channel);
+
             if (channel.Joined.Value)
             {
                 api.Queue(new LeaveChannelRequest(channel));
                 channel.Joined.Value = false;
             }
         }
 
+        /// <summary>
+        /// Opens the most recently closed channel that has not
+        /// already been reopened
+        /// Works similarly to reopening the last closed tab on a web browser.
+        /// </summary>
+        public void JoinLastClosedChannel()
+        {
+            if (closedChannels.Count <= 0)
+            {
+                return;
+            }
+
+            Channel lastClosedChannel = closedChannels.Last();
+
+            closedChannels.Remove(lastClosedChannel);
+
+            // If the user already joined the channel, try the next
+            // channel in the list
+            if (joinedChannels.IndexOf(lastClosedChannel) >= 0)
+            {
+                JoinLastClosedChannel();
```

Comment:
I'm not really comfortable with using recursion here, especially seeing that the worst case depth is 50 calls, and an iterative solution would work just fine.

As an aside, you could avoid this edge case whatsoever by checking if any new channel joined had been previously closed, and removing it from the list in that case. That would probably not play well with `LimitedCapacityQueue`, however, as it lacks a random access removal operation.

Reply:
Thanks for the feedback. I'll fix it to get rid of the recursion. As for avoiding the edge case, I thought about adding some deduplicate code. However, this code would have to be run whenever the user closed a tab, which could make closing a tab slow should the closedChannel list get large. I intentionally saved the checks for when the user actually used the Ctrl+Shift+t shortcut, this way the new feature would have minimal impact on the performance of any other part of the code.

Reply:
I mean I don't think it would be *that* much worse than the current code. Let's denote the number of previously-closed channels with `n`, and the number of currently-joined channels with `m`.

Current code:

* An item is added to `closedChannels` on every channel close, which is an amortised `O(1)` operation - amortised because the list has to be expanded in `O(n)` time every now and again, but as far as I know .NET uses the proper exponential reallocation scheme to achieve amortised `O(1)`. (As an aside, `closedhannels` is not currently checked for duplicates when added to, which seems unintended.)
* Each attempt to restore a channel is an `O(m)` operation (worst case for the `IndexOf()` to check if the channel we're trying to restore is open), and there can be up to `n` attempts, yielding a total of `O(n * m)`.

What I'm proposing:

* Adding to `closedChannels` would be the same as in the current code. (Preferably with duplicate checks.)
* When joining a channel, we'd have to check closed channels to make sure we clean them up, which is an `O(n)` operation. Joining `m` channels would therefore be `O(n * m)` again (but in `m` different operations instead of one).
* Restoring the last closed channel would be `O(1)` again, as the only thing to be done is to rejoin the last closed-but-not-reopened channel from `closedChannels`.

I guess the argument could be made that joining channels is a more frequent operation than restoring the last closed one, but I'm not sure `m` and `n` will ever get large enough where that will be a problem. That said I don't insist on this solution - I'm fine with just removing the recursion.